### PR TITLE
Add Fontconfig font search

### PR DESCRIPTION
**Pull request by asinghvi17**
# Description

This PR introduces a method to use Fontconfig to find and create font sets.

Basically, what this does is introduce 3 main methods.

1. The ability to obtain all styles in a font family (`populate_font_family`) - returns a NamedTuple of Symbol `styles => FTFont (Makie.NativeFont)`.
2. The ability to search for a particular style within a font family, along with some basic fallbacks (regular can be roman or book, for example, if there is no style called `regular`) - `search_font(family, style)`
3. The ability to cache font families and regularize some names, so that `(:regular, :bold, :italic, :bold_italic)` are always available, even if the font has obliques instead of italics.  Individual styles can still be referred to by their original names.  This is the `FontFamily` interface, represented by the `FontFamily` struct, which the user can edit.  It also automatically searches for a style if it can't exactly match, making it a bit more dynamic than what we currently have.

This results in the following:

```julia
julia> FontFamily("TeX Gyre Heros Makie")
FontFamily("TeX Gyre Heros Makie", Attributes with 5 entries:
  bold => FTFont (family = TeX Gyre Heros Makie, style = Bold)
  bold italic => FTFont (family = TeX Gyre Heros Makie, style = Bold Italic)
  bold_italic => FTFont (family = TeX Gyre Heros Makie, style = Bold Italic)
  italic => FTFont (family = TeX Gyre Heros Makie, style = Italic)
  regular => FTFont (family = TeX Gyre Heros Makie, style = Regular))
```
and `fonts = "TeX Gyre Heros Makie"` is equivalent to `fonts = FontFamily("TeX Gyre Heros Makie")`, which is what has been provided in the default theme.  This way, people can access the exact font style (`Extracondensed Semibold` or what have you) as well as some basic shortcuts.

The user interface is still not set in stone, and this is very much a draft PR.

## Type of change

Delete options that do not apply:

- [x] New feature (non-breaking change which adds functionality)

## TODOs

- [ ] Finalize the FontFamily interface and try to make it more efficient.
- [ ] Properly document the FontFamily interface - people shouldn't be passing strings and expecting elements of the family.
- [x] Add an entry in NEWS.md (for new features and breaking changes)
- [x] Add or change relevant sections in the documentation
- [ ] Add unit tests for new algorithms, conversion methods, etc.
- [ ] Add reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

**Comment by SimonDanisch** [comment]
I think most of the code, the pr + development of this feature should live in Fontconfig ;)

File: src/fonts.jl [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    fs = unsafe_load(fs_ptr)
+    patterns = Fontconfig.Pattern[]
+
+    for i in 1:fs.nfont
+        push!(patterns, Fontconfig.Pattern(unsafe_load(fs.fonts, i)))
+    end
+
+    ccall((:FcObjectSetDestroy, Fontconfig.libfontconfig), Nothing, (Ptr{Nothing},), os)
+    return patterns
+end
```
Comment (asinghvi17) [inline src/fonts.jl:28]: This can go, since it was merged into Fontconfig.

File: src/fonts.jl [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+Returns a String.
+"""
+function extract_fc_attr_int(pat::Fontconfig.Pattern, attr::String; num::Int = 0)
+       local output_int = Ref{Cint}()
+       success = @ccall Fontconfig.Fontconfig_jll.libfontconfig.FcPatternGetInteger(pat.ptr::Ptr{Cvoid}, attr::Cstring, num::Cint, output_int::Ptr{Cint})::Cint
+       if success != 0
+           error("The operation did not succeed!  Fontconfig returned the error code $success when extracting the $(num)th entry of attribute $attr from pattern $pat.")
+       end
+       return output_int[]
+end
```
Comment (asinghvi17) [inline src/fonts.jl:61]: These can be merged into Fontconfig

File: src/fonts.jl [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+function populate_font_family(family::String; additional_params_for_fontconfig...)
+    # Create a pattern to search for
+    pattern = Fontconfig.Pattern(; family = family, additional_params_for_fontconfig...)
+    # Use Fontconfig to generate a list of styles from that pattern.
+    font_list = fclist(pattern, ["family", "style", "file", "index"])
+    # Extract the style names (regular, light, bold, oblique, etc)
+    styles = extract_fc_attr_string.(font_list, "style")
+    # Return a NamedTuple
+    return (; (Symbol.(lowercase.(styles)) .=> ftfont_from_fc_pattern.(font_list))...)
+end
```
Comment (asinghvi17) [inline src/fonts.jl:103]: not too sure where this could go - either with FontFamily in Fontconfig, or here in Makie

File: src/fonts.jl [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+function is_bold(font::FreeTypeAbstraction.FTFont)
+    return font.style_flags & 1
+end
+
+function is_italic_or_oblique(font::FreeTypeAbstraction.FTFont)
+    return font.style_flags & 2
+end
+
+
+function search_font(family::String, style::String = "regular"; exact = false)::NativeFont
```
Comment (asinghvi17) [inline src/fonts.jl:117]: I could move a slightly refactored form of this to Fontconfig, where the user can pass various substitutions or required elements that they want.

File: src/fonts.jl [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    if any(isequal(style), key_s)
+        return list[findfirst(isequal(style), key_s)]
+    else
+        return list[1]
+    end
+
+end
+
+"""
+    FontFamily(family::String; search_kwargs)
```
Comment (asinghvi17) [inline src/fonts.jl:183]: From here on down it'll all be moved to Fontconfig

**Review by asinghvi17 — COMMENTED** [review]
I just used comments to outline a brief plan of what could go where.  Does that make sense to you @SimonDanisch?

**Comment by asinghvi17** [comment]
Closing in favour of a new PR

**Pull request closed without merging**